app/services/assignment_service.py

```diff
@@ -267,60 +289,78 @@ def _compute_master_notebook_name(assignment_name: str) -> str:
         return f"{ assignment_name }-prof.ipynb"
 
 class InstructorAssignmentService(AssignmentService):
-    def __init__(self, session: Session, instructor: InstructorModel, assignment: AssignmentModel):
+    def __init__(self, session: Session, instructor_model: InstructorModel, assignment_model: AssignmentModel, course_model: CourseModel):
         super().__init__(session)
-        self.instructor = instructor
-        self.assignment = assignment
+        self.instructor_model = instructor_model
+        self.assignment_model = assignment_model
+        self.course_model = course_model 
 
-    def get_is_available(self) -> bool:
-        if not self.assignment.is_created: return False
+    # The release date for a specific student, considering extra_time
+    def get_adjusted_available_date(self) -> datetime | None:
+        assignment_open_date = self.assignment_model.available_date or self.course_model.start_at
+        if assignment_open_date is None: return None
+        
+        return assignment_open_date
 
-        current_timestamp = self.session.scalar(func.current_timestamp())
-        return current_timestamp >= self.assignment.available_date
-    
-    def get_is_closed(self) -> bool:
-        if not self.assignment.is_created: return False
+    # The due date for a specific student, considering extra_time
+    def get_adjusted_due_date(self) -> datetime | None:
+        assignment_due_date = self.assignment_model.due_date or self.course_model.end_at
+        if assignment_due_date is None: return None
+
+        return assignment_due_date
+
+    def get_assignment_status(self) -> AssignmentStatus:
+        if not self.assignment_model.is_published: return AssignmentStatus.UNPUBLISHED
 
         current_timestamp = self.session.scalar(func.current_timestamp())
-        return current_timestamp > self.assignment.due_date
+        adjusted_available_date = self.get_adjusted_available_date()
+        adjusted_due_date = self.get_adjusted_due_date()
+        
+        if adjusted_available_date is None or adjusted_available_date >= current_timestamp: return AssignmentStatus.UPCOMING
+
+        if adjusted_due_date and adjusted_due_date > current_timestamp: return AssignmentStatus.OPEN
```

Comment:
This logic is still incorrect. `IF adjusted_available_date is None`, the assignment has no open date, it is always open. That means it should not be considered upcoming. Perhaps some variables for these would be helpful.

`IF adjusted_due_date` <=> `IF adjusted_due_date is not None`, you are checking the assignment has a due date? which is confusing?

Reply:
The reason I chose to set the assignment to UPCOMING if there is no available_date _or_ course start date set and CLOSED if no due_date _or_ course end date is set is because I don't think it's a good idea for an assignment to accidentally be set to OPEN just because the class isn't configured fully. And there's an easy way to fix each one: set the course start and end dates. That will fix it for all assignments. What you're proposing is to set it to OPEN for both of these scenarios, which I think is a worse option because if the professor hasn't set a course start date and then assignment 2 comes around and they create an assignment without an available date, it will automatically just be set to OPEN right then.

As for the comments within the method, that's a good point, I can do that. And regarding

Reply:
I agree, to some extent. But it's a professor's responsibility to understand how Canvas works. If they create and publish an assignment with no open/due date, then Canvas treats that as an assignment that is always open.

I'm not sure I follow what you're saying initially. Even if the course isn't misconfigured and does have a start/end date set, an assignment without an open/close date would still be automatically open for the entirety of the course. That's effectively the same as just always open anyways.

As an aside: consider a course without start/end dates configured properly, and consider a published assignment with an available date in the past but no due date set. It won't be marked UPCOMING since available date has already past, but due_date is None so it's going to return that the assignment is closed. That doesn't make sense.

If you really want to disallow that sort of behavior, then just always return UPCOMING for everything until both the course start and due date are set. Then implement the ideal logic instead